llvm/lib/Target/AIE/AIE2InstrInfo.cpp

```diff
@@ -1305,3 +1305,57 @@ bool AIE2InstrInfo::canHoistCheapInst(const MachineInstr &MI) const {
          AIE2::eSRegClass.hasSubClassEq(
              MRI.getRegClass(MI.getOperand(0).getReg()));
 }
+
+bool AIE2InstrInfo::combineLoHiMI(MachineInstr &MI,
+                                  MachineRegisterInfo &MRI) const {
```

Comment:
I feel a lot of the machinery here could be share between AIE versions. Could you maybe move the MI creation/manipulation to the PostSelectOptimize pass? Then I would suggest only implementing an interface like below for each AIEInstrInfo version. E.g. for AIE2:

```
struct LoHiCombineInfo {
  unsigned CombinedOpcode;
  unsigned OpIdxToAggregate;
};

std::optional<LoHiCombineInfo> AIE2InstrInfo::getCombinedLoHiOpcode(LoMI, HiMI) {
  unsigned LoOpcode = LoMI.getOpcode();
  unsigned HIOpcode = HiMI.getOpcode();
  if (LoOpcode != HiOpcode)
    return {};

  switch (LoOpcode) {
  case AIE2::VUPS_S32_D16_mv_ups_w2b:
    if (LoMI.getOperand(2) == HiMI.getOperand(2))
      return LoHiCombineInfo{VUPS_S32_D16_mv_ups_x2c, /*OpIdxToAggregate=*/1}
    break;
  case ...
  }
  return {}
}

Reply:
Maybe `LoHiCombineInfo` could also include register classes for input and output.....